### PR TITLE
BUG: Fix all kinds of problems when itemsize == 0

### PR DESCRIPTION
**Pull request by eric-wieser**
Split off all of the bugfixes from #8970.

### CODE REVIEW

File: numpy/core/src/multiarray/ctors.c [hunk 1]
```diff
@@ -3711,14 +3711,15 @@ PyArray_FromIter(PyObject *obj, PyArray_Descr *dtype, npy_intp count)
...
             /*
               Grow PyArray_DATA(ret):
               this is similar for the strategy for PyListObject, but we use
               50% overallocation => 0, 4, 8, 14, 23, 36, 56, 86 ...
             */
             elcount = (i >> 1) + (i < 4 ? 4 : 2) + i;
-            if (elcount <= NPY_MAX_INTP/elsize) {
-                new_data = PyDataMem_RENEW(PyArray_DATA(ret), elcount * elsize);
+            if (!npy_mul_with_overflow_intp(&nbytes, elcount, elsize)) {
+                new_data = PyDataMem_RENEW(PyArray_DATA(ret), nbytes);
```
Comment (juliantaylor) [inline numpy/core/src/multiarray/ctors.c]: the else clause and the `new_data` variable can be removed

File: numpy/core/src/multiarray/shape.c [hunk 2]
```diff
@@ -126,22 +121,21 @@ PyArray_Resize(PyArrayObject *self, PyArray_Dims *newshape, int refcheck,
...
+    if ((newnbytes > oldnbytes) && PyArray_ISWRITEABLE(self)) {
         /* Fill new memory with zeros */
-        elsize = PyArray_DESCR(self)->elsize;
         if (PyDataType_FLAGCHK(PyArray_DESCR(self), NPY_ITEM_REFCOUNT)) {
             PyObject *zero = PyInt_FromLong(0);
             char *optr;
-            optr = PyArray_BYTES(self) + oldsize*elsize;
-            n = newsize - oldsize;
+            optr = PyArray_BYTES(self) + oldnbytes;
+            n = (newnbytes - oldnbytes) / elsize;
```
Comment (juliantaylor) [inline numpy/core/src/multiarray/shape.c]: this looks scetchy as the == 0 check is removed.
It might not go in here as then newbytes and oldnbytes are 0 but it's still dangerous

Comment (eric-wieser) [inline numpy/core/src/multiarray/shape.c]: The idea here is that if the number of bytes are different, the element must be non empty. I guess rather than doing the division, we could just calculate this directly though

File: numpy/core/src/multiarray/ctors.c (hunk 1 above)
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: What do you mean by removing the variable?

Fixing that feels kinda out of scope for this patch

File: numpy/core/src/multiarray/getset.c [hunk 3]
```diff
@@ -472,25 +472,19 @@ array_descr_set(PyArrayObject *self, PyObject *arg)
...
-        else {
-            PyErr_SetString(PyExc_TypeError,
-                    "data-type must not be 0-sized");
-            Py_DECREF(newtype);
+    /*
+     * Treat V0 as resizable void - unless the destination is already V0, then
+     * don't allow np.void to be duplicated
+     */
+    if (newtype->type_num == NPY_VOID && newtype->elsize == 0
+                                      && PyArray_DESCR(self)->elsize != 0) {
```
Comment (juliantaylor) [inline numpy/core/src/multiarray/getset.c]: our usual style is having the operator on the line before but indenting the next line

File: numpy/core/src/multiarray/shape.c (hunk 2 above)
Comment (juliantaylor) [inline numpy/core/src/multiarray/shape.c]: hm its probably good as is.

File: numpy/core/src/multiarray/getset.c (hunk 3 above)
Comment (eric-wieser) [inline numpy/core/src/multiarray/getset.c]: That doesn't match with what I remember seeing, eg in `mapping.c`. Operators should be at the start of the line, so that they're not missed when reading through.

I can wrap after `NPY_VOID` if you'd like, but I'm not sure what level to indent to as a result

File: numpy/core/src/multiarray/shape.c (hunk 2 above)
Comment (eric-wieser) [inline numpy/core/src/multiarray/shape.c]: Patch incoming, I think it makes it cleaner

File: numpy/core/src/multiarray/getset.c (hunk 3 above)
Comment (charris) [inline numpy/core/src/multiarray/getset.c]: The usual style (Python recommended) is trailing operators and two indents on the the next line too separate the boolean expression from the body.

Comment (eric-wieser) [inline numpy/core/src/multiarray/getset.c]: Is there something like CPEP 8 for cpython style guidelines?

Comment (MSeifert04) [inline numpy/core/src/multiarray/getset.c]: You mean [PEP7](https://www.python.org/dev/peps/pep-0007/)?

File: numpy/core/src/multiarray/shape.c (hunk 2 above)
Comment (eric-wieser) [inline numpy/core/src/multiarray/shape.c]: Also uncovered another unrelated bug

File: numpy/core/src/multiarray/getset.c (hunk 3 above)
Comment (charris) [inline numpy/core/src/multiarray/getset.c]: Also `doc/C_STYLE_GUIDE.rst.txt`. Here, second line just two indents from the `if`.

Comment (eric-wieser) [inline numpy/core/src/multiarray/getset.c]: Already fixed here, I think

File: numpy/core/src/multiarray/ctors.c (hunk 1 above)
Comment (charris) [inline numpy/core/src/multiarray/ctors.c]: IIRC, realloc will free the memory if size is zero, which we don't want to do. Worth checking, normally we keep a minimum of 1 byte to avoid problems.

Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: Sometimes, yes - but we don't actually care. All that matters is if `realloc` returns NULL, (which `realloc(..., 0)` does on some platforms, in which case we fail with a memory error.

Either way, we don't need to worry about this yet - this function has a check for elsize == 0, which we shouldn't remove unless #8970 is merged.

So yes, technically, this file didn't need a patch at all - but I'd argue that `npy_mul_with_overflow_intp` does a better job of communicating the intention

**Comment by eric-wieser** [comment]
Rebased to solve conflicts.

File: numpy/core/src/multiarray/descriptor.c [hunk 4]
```diff
@@ -348,14 +350,21 @@ _convert_from_tuple(PyObject *obj)
...
+        if (overflowed) {
             PyErr_SetString(PyExc_ValueError,
                             "invalid shape in fixed-type tuple: dtype size in "
                             "bytes must fit into a C int.");
             npy_free_cache_dim_obj(shape);
             goto fail;
         }
-        newdescr->elsize = type->elsize * items;
+        newdescr->elsize = nbytes;
         if (newdescr->elsize == -1) {
```
Comment (charris) [inline numpy/core/src/multiarray/descriptor.c]: Can `nbytes` be -1? If so, could use a note. I'm guessing a single item of size 0, which seems restrictive.

File: numpy/core/src/multiarray/getset.c [hunk 5]
```diff
@@ -469,22 +469,18 @@ array_descr_set(PyArrayObject *self, PyObject *arg)
...
-            newtype->elsize = PyArray_DESCR(self)->elsize;
-        }
-        /* But no other flexible types */
-        else {
-            PyErr_SetString(PyExc_TypeError,
-                    "data-type must not be 0-sized");
-            Py_DECREF(newtype);
+    /*
+     * Treat V0 as resizable void - unless the destination is already V0, then
+     * don't allow np.void to be duplicated
```
Comment (charris) [inline numpy/core/src/multiarray/getset.c]: Why not?

File: numpy/core/src/multiarray/methods.c [hunk 6]
```diff
@@ -1712,11 +1715,9 @@ array_setstate(PyArrayObject *self, PyObject *args)
         return NULL;
     }
     size = PyArray_MultiplyList(dimensions, nd);
-    if (PyArray_DESCR(self)->elsize == 0) {
-        PyErr_SetString(PyExc_ValueError, "Invalid data-type size.");
-        return NULL;
-    }
-    if (size < 0 || size > NPY_MAX_INTP / PyArray_DESCR(self)->elsize) {
+    overflowed = npy_mul_with_overflow_intp(
+        &nbytes, size, PyArray_DESCR(self)->elsize);
```
Comment (charris) [inline numpy/core/src/multiarray/methods.c]: Doesn't `size` need to be >= 0 in order to use `npy_mul_with_overflow_intp`? Seems it should be checked first to avoid a function call.

Comment (charris) [reply numpy/core/src/multiarray/methods.c]: Could do something like
```
invalid = size < 0 || npy_mul_with_overflow_intp(...)
```

File: numpy/core/src/multiarray/methods.c [hunk 7]
```diff
@@ -1822,7 +1823,7 @@ array_setstate(PyArrayObject *self, PyObject *args)
             }
             if (swap) {
                 /* byte-swap on pickle-read */
-                npy_intp numels = num / PyArray_DESCR(self)->elsize;
+                npy_intp numels = PyArray_SIZE(self);
```
Comment (charris) [inline numpy/core/src/multiarray/methods.c]: Seems sensible, but why was it done differently before?

File: numpy/core/src/multiarray/getset.c (hunk 5 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/getset.c]: For the same reason that `np.dtype(np.void) is np.dtype(np.void)` is desirable. Without this special case, `v0_arr.view(np.void).dtype is not np.void`.

File: numpy/core/src/multiarray/descriptor.c (hunk 4 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/descriptor.c]: Huh, that's pretty weird. This couldn't happen before either, could it?

File: numpy/core/src/multiarray/shape.c [hunk 8]
```diff
@@ -111,11 +112,11 @@ PyArray_Resize(PyArrayObject *self, PyArray_Dims *newshape, int refcheck,
             return NULL;
         }
 
-        if (newsize == 0) {
-            sd = PyArray_DESCR(self)->elsize;
+        if (newnbytes == 0) {
```
Comment (charris) [inline numpy/core/src/multiarray/shape.c]: Could do
```
sd = newnbytes ? newnbytes : elsize
```

File: numpy/core/src/multiarray/methods.c (hunk 6 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/methods.c]: No need, there's already a explicit `a != 0` case in that function.

There's no point avoiding the function call, because optimizing the performance of empty arrays is pretty pointless

Comment (eric-wieser) [reply numpy/core/src/multiarray/methods.c]: Oh I see, the issue is size == -1. Maybe I'll just duplicate the below `if`.

File: numpy/core/src/multiarray/methods.c (hunk 7 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/methods.c]: To avoid recalculating `num`, most likely

File: numpy/core/src/multiarray/shape.c [hunk 9]
```diff
@@ -142,7 +142,7 @@ PyArray_Resize(PyArrayObject *self, PyArray_Dims *newshape, int refcheck,
             Py_DECREF(zero);
         }
         else{
-            memset(PyArray_BYTES(self)+oldsize*elsize, 0, (newsize-oldsize)*elsize);
+            memset(PyArray_BYTES(self) + oldnbytes, 0, newnbytes - oldnbytes);
```
Comment (charris) [inline numpy/core/src/multiarray/shape.c]: `newnbytes - oldnbytes` >= 0?

File: numpy/core/src/multiarray/shape.c (hunk 8 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/shape.c]: Would prefer to keep the `== 0`, because that's handling the special meaning of `0` which is "allow resizing".

Could flatten the `if` though

File: numpy/core/src/multiarray/shape.c (hunk 9 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/shape.c]: Yep, guaranteed by the enclosing `if`

File: numpy/core/tests/test_multiarray.py [hunk 10]
```diff
@@ -1201,6 +1201,74 @@ def test_count_nonzero_unaligned(self):
             assert_equal(np.count_nonzero(a), builtins.sum(a.tolist()))
 
 
+class TestZeroSizeFlexible(object):
+    @staticmethod
```
Comment (charris) [inline numpy/core/tests/test_multiarray.py]: Is there an advantage to making this a staticmethod?

Comment (eric-wieser) [reply numpy/core/tests/test_multiarray.py]: To document that it doesn't use `self`?

**Comment by charris** [comment]
Looks pretty good, just a couple of questions.

**Comment by charris** [comment]
A release note under "Improvements" would be helpful. Someone might want to know just when the new functionality became available.

File: numpy/core/tests/test_multiarray.py (hunk 10 above)
Comment (charris) [reply numpy/core/tests/test_multiarray.py]: Well, it does that, don't know if it is worth it as the normal impulse would be to use it only if needed. Then again, my Python experience started when Python 21 was the newest and greatest, so programming conventions may have changed. In fact, it is the latter possibility that interests me. Do you know if it is the case that this sort of programming style is being taught or encouraged?

Comment (eric-wieser) [reply numpy/core/tests/test_multiarray.py]: `TestCase` objects (or the inferred use of them by pytest) are pretty weird anyway, because an new one is instantiated for each method call by the test runner. I can change this if you feel strongly.

Comment (charris) [reply numpy/core/tests/test_multiarray.py]: I don't feel strongly about it. On occasion I find your coding style a bit unusual, but not wrong. Just trying to figure out if it is a personal style, or the future of python coding ;)

Comment (eric-wieser) [reply numpy/core/tests/test_multiarray.py]: Could easily be personal style. I guess it probably comes from my java background, but even there it might just have been personal style.

**Comment by eric-wieser** [comment]
Caught a memory leak on the error path here too

**Comment by charris** [comment]
Thanks Eric.
